[CI] ReindexFailureTests testAbortOnVersionConflict, testFailuresCauseAbortDefault, testResponseOnSearchFailure

https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.x+multijob-unix-compatibility/os=centos/479/console

[build-479-ReindexFailureTests.txt](https://github.com/elastic/elasticsearch/files/1598563/build-479-ReindexFailureTests.txt)

```
ERROR   51.4s J3 | ReindexFailureTests.testResponseOnSearchFailure <<< FAILURES!
   > Throwable #1: java.lang.AssertionError:
   > Expected: an empty iterable
   > dices:data/write/bulk[s]]]; nested: IndexNotFoundException[no such index];>,<RemoteTransportException[[node_sd3][127.0.0.1:35162][indices:data/write/bulk[s]]]; nested: IndexNotFoundException[no such index];>,<RemoteTransportException[[node_sd3][127.0.0.1:35162][indices:data/write/bulk[s]]]; nested: IndexNotFoundException[no such index];>,<RemoteTransportException[[node_sd3][127.0.0.1:35162][indices:data/write/bulk[s]]]; nested: IndexNotFoundException[no such index];>,<RemoteTransportException[[node_sd3][127.0.0.1:35162][indices:data/write/bulk[s]]]; nested: IndexNotFoundException[no such index];>,<RemoteTransportException[[node_sd3][127.0.0.1:35162][indices:data/write/bulk[s]]]; nested: IndexNotFoundException[no such index];>,<RemoteTransportException[[node_sd3][127.0.0.1:35162][indices:data/write/bulk[s]]]; nested: IndexNotFoundException[no such index];>,<RemoteTransportException[[node_sd3][127.0.0.1:35162][indices:data/write/bulk[s]]]; nested: IndexNotFoundException[no such index];>,<RemoteTransportException[[node_sd1][127.0.0.1:43434][indices:data/write/index]]; nested: RemoteTransportException[[node_sd3][127.0.0.1:35162][indices:data/write/bulk[s]]]; nested: IndexNotFoundException[no such index];>,<RemoteTransportException[[node_sd2][127.0.0.1:35287][indices:data/write/bulk[s]]]; nested: IndexNotFoundException[no such index];>,<RemoteTransportException[[node_sd1][127.0.0.1:43434][indices:data/write/index]]; nested: RemoteTransportException[[node_sd3][127.0.0.1:35162][indices:data/write/bulk[s]]]; nested: IndexNotFoundException[no such index];>,<RemoteTransportException[[node_sd2][127.0.0.1:35287][indices:data/write/bulk[s]]]; nested: IndexNotFoundException[no such index];>,<RemoteTransportException[[node_sd2][127.0.0.1:35287][indices:data/write/bulk[s]]]; nested: IndexNotFoundException[no such index];>,<RemoteTransportException[[node_sd2][127.0.0.1:35287][indices:data/write/bulk[s]]]; nested: IndexNotFoundException[no such index];>,<RemoteTransportException[[node_sd2][127.0.0.1:35287][indices:data/write/bulk[s]]]; nested: IndexNotFoundException[no such index];>,<RemoteTransportException[[node_sd2][127.0.0.1:35287][indices:data/write/bulk[s]]]; nested: IndexNotFoundException[no such index];>,<RemoteTransportException[[node_sd2][127.0.0.1:35287][indices:data/write/bulk[s]]]; nested: IndexNotFoundException[no such index];>,<RemoteTransportException[[node_sd1][127.0.0.1:43434][indices:data/write/index]]; nested: RemoteTransportException[[node_sd2][127.0.0.1:35287][indices:data/write/bulk[s]]]; nested: IndexNotFoundException[no such index];>,<RemoteTransportException[[node_sd2][127.0.0.1:35287][indices:data/write/bulk[s]]]; nested: IndexNotFoundException[no such index];>,<RemoteTransportException[[node_sd2][127.0.0.1:35287][indices:data/write/bulk[s]]]; nested: IndexNotFoundException[no such index];>,<[source/sj2mv1B6SP6OUxseTSjeOQ] IndexNotFoundException[no such index]>,<[source/sj2mv1B6SP6OUxseTSjeOQ] IndexNotFoundException[no such index]>,<[source/sj2mv1B6SP6OUxseTSjeOQ] IndexNotFoundException[no such index]>,<[source/sj2mv1B6SP6OUxseTSjeOQ] IndexNotFoundException[no such index]>,<RemoteTransportException[[node_sd1][127.0.0.1:43434][indices:data/write/index]]; nested: IndexNotFoundException[no such index];>,<[source/sj2mv1B6SP6OUxseTSjeOQ] IndexNotFoundException[no such index]>,<[source/sj2mv1B6SP6OUxseTSjeOQ] IndexNotFoundException[no such index]>,<[source/sj2mv1B6SP6OUxseTSjeOQ] IndexNotFoundException[no such index]>,<[source/sj2mv1B6SP6OUxseTSjeOQ] IndexNotFoundException[no such index]>,<[source/sj2mv1B6SP6OUxseTSjeOQ] IndexNotFoundException[no such index]>,<[source/sj2mv1B6SP6OUxseTSjeOQ] IndexNotFoundException[no such index]>,<[source/sj2mv1B6SP6OUxseTSjeOQ] IndexNotFoundException[no such index]>,<[source/sj2mv1B6SP6OUxseTSjeOQ] IndexNotFoundException[no such index]>,<[source/sj2mv1B6SP6OUxseTSjeOQ] IndexNotFoundException[no such index]>,<[source/sj2mv1B6SP6OUxseTSjeOQ] IndexNotFounption[no such index]>,<[source/sj2mv1B6SP6OUxseTSjeOQ] IndexNotFoundException[no such index]>,<[source/sj2mv1B6SP6OUxseTSjeOQ] IndexNotFoundException[no such index]>,<[source/sj2mv1B6SP6OUxseTSjeOQ] IndexNotFoundException[no such index]>]
   >    at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
   >    at org.elasticsearch.test.ESIntegTestCase.indexRandom(ESIntegTestCase.java:1451)
   >    at org.elasticsearch.test.ESIntegTestCase.indexRandom(ESIntegTestCase.java:1369)
   >    at org.elasticsearch.test.ESIntegTestCase.indexRandom(ESIntegTestCase.java:1353)
   >    at org.elasticsearch.index.reindex.ReindexFailureTests.indexDocs(ReindexFailureTests.java:128)
   >    at org.elasticsearch.index.reindex.ReindexFailureTests.testResponseOnSearchFailure(ReindexFailureTests.java:100)
   >    at java.lang.Thread.run(Thread.java:748)Throwable #2: java.lang.AssertionError: There are still [1] in-flight contexts. The first one's creation site is listed as the cause of this exception.
   >    at org.elasticsearch.search.MockSearchService.assertNoInFlightContext(MockSearchService.java:51)
   >    at org.elasticsearch.test.ESTestCase.lambda$ensureAllSearchContextsReleased$4(ESTestCase.java:421)
   >    at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:731)
   >    at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:705)
   >    at org.elasticsearch.test.ESTestCase.ensureAllSearchContextsReleased(ESTestCase.java:421)
   >    at org.elasticsearch.test.ESIntegTestCase.cleanUpCluster(ESIntegTestCase.java:2072)
   >    at java.lang.Thread.run(Thread.java:748)
   >    Suppressed: java.lang.AssertionError: There are still [1] in-flight contexts. The first one's creation site is listed as the cause of this exception.
   >            at org.elasticsearch.search.MockSearchService.assertNoInFlightContext(MockSearchService.java:51)
   >            at org.elasticsearch.test.ESTestCase.lambda$ensureAllSearchContextsReleased$4(ESTestCase.java:421)
   >            at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:719)
   >            ... 37 more
```

```
gradle :modules:reindex:test \
  -Dtests.seed=401B0C5B3206920F \
  -Dtests.class=org.elasticsearch.index.reindex.ReindexFailureTests \
  -Dtests.method="testAbortOnVersionConflict" \
  -Dtests.security.manager=true \
  -Dtests.locale=sr-Latn \
  -Dtests.timezone=America/Cuiaba
```

Another instance on master: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+intake/3712/

```
java.lang.AssertionError: There are still [1] in-flight contexts. The first one's creation site is listed as the cause of this exception.
	at org.elasticsearch.search.MockSearchService.assertNoInFlightContext(MockSearchService.java:51)
	at org.elasticsearch.test.ESTestCase.lambda$ensureAllSearchContextsReleased$6(ESTestCase.java:503)
	at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:862)
	at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:836)
	at org.elasticsearch.test.ESTestCase.ensureAllSearchContextsReleased(ESTestCase.java:503)
```
```
REPRODUCE WITH: ./gradlew :modules:reindex:test --tests "org.elasticsearch.index.reindex.ReindexFailureTests.testResponseOnSearchFailure" -Dtests.seed=25A63C7B43796EB8 -Dtests.security.manager=true -Dtests.locale=az -Dtests.timezone=Africa/Ceuta -Dcompiler.java=12 -Druntime.java=11
```

Log: [reindex.txt](https://github.com/elastic/elasticsearch/files/3217741/reindex.txt)

--- a/server/src/main/java/org/elasticsearch/search/SearchService.java
+++ b/server/src/main/java/org/elasticsearch/search/SearchService.java
@@ -594,19 +594,35 @@ public class SearchService extends AbstractLifecycleComponent implements IndexEv
         }
 
         SearchContext context = createContext(request);
+        onNewContext(context);
         boolean success = false;
         try {
             putContext(context);
-            if (request.scroll() != null) {
+            success = true;
+            return context;
+        } finally {
+            if (success == false) {
+                freeContext(context.id());
+            }
+        }
+    }
+
+    private void onNewContext(SearchContext context) {
+        boolean success = false;
+        try {
+            if (context.scrollContext() != null) {
                 openScrollContexts.incrementAndGet();
                 context.indexShard().getSearchOperationListener().onNewScrollContext(context);
             }
             context.indexShard().getSearchOperationListener().onNewContext(context);
             success = true;
-            return context;
         } finally {
-            if (!success) {
-                freeContext(context.id());
+            // currently, the concrete listener is CompositeListener, which swallows exceptions, but here we anyway try to do the
+            // right thing by closing and notifying onFreeXXX in case one of the listeners fails with an exception in the future.
+            if (success == false) {
+                try (SearchContext dummy = context) {
+                    onFreeContext(context);
+                }
             }
         }
     }
@@ -693,19 +709,24 @@ public class SearchService extends AbstractLifecycleComponent implements IndexEv
     public boolean freeContext(long id) {
         final SearchContext context = removeContext(id);
         if (context != null) {
-            assert context.refCount() > 0 : " refCount must be > 0: " + context.refCount();
             try {
-                context.indexShard().getSearchOperationListener().onFreeContext(context);
-                if (context.scrollContext() != null) {
-                    openScrollContexts.decrementAndGet();
-                    context.indexShard().getSearchOperationListener().onFreeScrollContext(context);
-                }
+                onFreeContext(context);
             } finally {
                 context.close();
             }
             return true;
         }
         return false;
+    }
+
+    private void onFreeContext(SearchContext context) {
+        assert context.refCount() > 0 : " refCount must be > 0: " + context.refCount();
+        assert activeContexts.containsKey(context.id()) == false;
+        context.indexShard().getSearchOperationListener().onFreeContext(context);
+        if (context.scrollContext() != null) {
+            openScrollContexts.decrementAndGet();
+            context.indexShard().getSearchOperationListener().onFreeScrollContext(context);
+        }
     }
 
     public void freeAllScrollContexts() {

--- a/server/src/test/java/org/elasticsearch/search/SearchServiceTests.java
+++ b/server/src/test/java/org/elasticsearch/search/SearchServiceTests.java
@@ -19,7 +19,6 @@
 package org.elasticsearch.search;
 
 import com.carrotsearch.hppc.IntArrayList;
-
 import org.apache.lucene.search.Query;
 import org.apache.lucene.store.AlreadyClosedException;
 import org.elasticsearch.ElasticsearchException;
@@ -52,6 +51,7 @@ import org.elasticsearch.index.query.QueryBuilder;
 import org.elasticsearch.index.query.QueryRewriteContext;
 import org.elasticsearch.index.query.QueryShardContext;
 import org.elasticsearch.index.query.TermQueryBuilder;
+import org.elasticsearch.index.search.stats.SearchStats;
 import org.elasticsearch.index.shard.IndexShard;
 import org.elasticsearch.index.shard.SearchOperationListener;
 import org.elasticsearch.index.shard.ShardId;
@@ -229,6 +229,7 @@ public class SearchServiceTests extends ESSingleNodeTestCase {
         AtomicBoolean running = new AtomicBoolean(true);
         CountDownLatch startGun = new CountDownLatch(1);
         Semaphore semaphore = new Semaphore(Integer.MAX_VALUE);
+
         final Thread thread = new Thread() {
             @Override
             public void run() {
@@ -267,10 +268,17 @@ public class SearchServiceTests extends ESSingleNodeTestCase {
                 try {
                     try {
                         PlainActionFuture<SearchPhaseResult> result = new PlainActionFuture<>();
-                        service.executeQueryPhase(
-                            new ShardSearchLocalRequest(indexShard.shardId(), 1, SearchType.DEFAULT,
+                        final boolean useScroll = randomBoolean();
+                        ShardSearchLocalRequest shardRequest;
+                        if (useScroll) {
+                            shardRequest = new ShardScrollRequestTest(indexShard.shardId());
+                        } else {
+                            shardRequest = new ShardSearchLocalRequest(indexShard.shardId(), 1, SearchType.DEFAULT,
                                 new SearchSourceBuilder(), new String[0], false, new AliasFilter(null, Strings.EMPTY_ARRAY), 1.0f,
-                                true, null, null),
+                                true, null, null);
+                        }
+                        service.executeQueryPhase(
+                            shardRequest,
                             new SearchTask(123L, "", "", "", null, Collections.emptyMap()), result);
                         SearchPhaseResult searchPhaseResult = result.get();
                         IntArrayList intCursors = new IntArrayList(1);
@@ -279,6 +287,9 @@ public class SearchServiceTests extends ESSingleNodeTestCase {
                         PlainActionFuture<FetchSearchResult> listener = new PlainActionFuture<>();
                         service.executeFetchPhase(req, new SearchTask(123L, "", "", "", null, Collections.emptyMap()), listener);
                         listener.get();
+                        if (useScroll) {
+                            service.freeContext(searchPhaseResult.getRequestId());
+                        }
                     } catch (ExecutionException ex) {
                         assertThat(ex.getCause(), instanceOf(RuntimeException.class));
                         throw ((RuntimeException)ex.getCause());
@@ -296,6 +307,13 @@ public class SearchServiceTests extends ESSingleNodeTestCase {
             thread.join();
             semaphore.acquire(Integer.MAX_VALUE);
         }
+
+        assertEquals(0, service.getActiveContexts());
+
+        SearchStats.Stats totalStats = indexShard.searchStats().getTotal();
+        assertEquals(0, totalStats.getQueryCurrent());
+        assertEquals(0, totalStats.getScrollCurrent());
+        assertEquals(0, totalStats.getFetchCurrent());
     }
 
     public void testTimeout() throws IOException {

